### ISSUE TITLE
Model Failed to Predict the Correct Class using One-Class

### ISSUE THREAD

**Issue by Avv22** [ultralytics/ultralytics#13597]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello Glenn Jocher,

I followed your steps to train the model on one class only. The objective was to detect based on colors of object parts and how many parts there in one image it the parts are group correctly. I labelled the dataset on one class as you see in data.yaml bleow:

```
train: train/images
val: test/images
test: valid/images

nc: 1
names: ['correct']

roboflow:
  workspace: x-k4a2j
  project: wiremismatch2
  version: 1
  license: CC BY 4.0
  url: https://universe.roboflow.com/x-k4a2j/wiremismatch2/dataset/1
```

For training, I run it as follows:

```
import torch
from ultralytics import YOLO

def train_model():
    # Check if CUDA is available and set the device accordingly
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    # Load the YOLOv8 model and move it to the GPU if available
    model = YOLO('yolov8n.pt')
    model.to(device)

    # Train the model
    results = model.train(data="data.yaml", epochs=100, imgsz=640, single_cls=True)

if __name__ == '__main__':
    train_model()
```

Original input image is 1080x1920 and the training size is 360x640. I did not get any single detection :( The object is not very small but a little bit small and has 4 parts crossing each other so it's a little bit occluded. On the dataset I did some preprocessing including increase hue and contrast a little bit to increase the color of some parts of the objects that's it. Also I did some rotation for the whole image not on the bounding box. Specifically, I augmented images to a hue between 10 to 20 for hue and 20 to 30 for saturation.

So basically, I was trying to let the model knows how object in it's natural shape only as I don't have the abnormal shape thus the usage of one class. It performs good on the video I trained it one originally (i mean the frames from the video I extracted for training) but not on test data! I have 1206 training images after augmentation, 169 validation images also after augmentation and 90 images testing.

Results of training as follows:

![image](https://github.com/ultralytics/ultralytics/assets/45768058/eba155e8-227d-4405-ac44-213703393dfb)

Labels:

![image](https://github.com/ultralytics/ultralytics/assets/45768058/64b52d6c-88ce-4b8b-affd-e06a742d2654)

Confusion matrix:

![image](https://github.com/ultralytics/ultralytics/assets/45768058/4490506f-aac0-45cd-888d-61d1447e830b)

Any suggestion please?

**Comment by glenn-jocher** [comment]
@Avv22 hello,

Thank you for providing detailed information about your issue. Let's work through this together to identify the root cause and find a solution.

### Steps to Verify and Troubleshoot

1. **Reproducible Example**:
   It looks like you've provided a good amount of detail, but to ensure we can reproduce the issue accurately, could you please share a minimum reproducible example? This includes a small subset of your dataset and the exact code you're using. You can refer to our guide on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

2. **Package Versions**:
   Ensure you are using the latest versions of `torch` and `ultralytics`. You can update your packages using the following commands:
   ```bash
   pip install --upgrade torch ultralytics
   ```
   After updating, please try running your training script again to see if the issue persists.

3. **Training Configuration**:
   Your training script looks mostly correct, but let's ensure a few things:
   - Verify that your dataset is correctly labeled and the `data.yaml` file paths are accurate.
   - Ensure that the `single_cls=True` argument is correctly applied in your training command.

4. **Model Performance**:
   Given that your model performs well on the training data but not on the test data, this might indicate overfitting. Here are a few suggestions to mitigate this:
   - **Increase Dataset Size**: If possible, add more diverse examples to your training dataset.
   - **Data Augmentation**: Continue using data augmentation techniques, but ensure they are applied consistently. Consider using additional augmentations like random cropping, flipping, and scaling.
   - **Regularization**: Implement techniques such as dropout or weight decay to help generalize better.

5. **Evaluation Metrics**:
   Monitor additional metrics such as precision, recall, and mAP during training to get a better understanding of your model's performance. Tools like TensorBoard or wandb can be very helpful for this.

### Example Code for Training with Augmentation
Here's a refined example of how you might set up your training script with additional augmentations:
```python
import torch
from ultralytics import YOLO

def train_model():
    # Check if CUDA is available and set the device accordingly
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    # Load the YOLOv8 model and move it to the GPU if available
    model = YOLO('yolov8n.pt')
    model.to(device)

    # Train the model with additional augmentations
    results = model.train(
        data="data.yaml",
        epochs=100,
        imgsz=640,
        single_cls=True,
        augment=True,  # Enable data augmentation
        hsv_h=0.015,  # Adjust hue
        hsv_s=0.7,    # Adjust saturation
        hsv_v=0.4,    # Adjust value
        degrees=10.0, # Rotate images
        translate=0.1, # Translate images
        scale=0.5,    # Scale images
        shear=0.0,    # Shear images
        flipud=0.0,   # Flip images vertically
        fliplr=0.5    # Flip images horizontally
    )

if __name__ == '__main__':
    train_model()
```

### Next Steps
- Please update your packages and try the refined training script.
- Share a minimum reproducible example if the issue persists.
- Monitor additional metrics during training to gain more insights.

Feel free to reach out if you have any further questions or need additional assistance. We're here to help! 😊

**Comment by Avv22** [comment]
> @Avv22 hello,
> 
> Thank you for providing detailed information about your issue. Let's work through this together to identify the root cause and find a solution.
> 
> ### Steps to Verify and Troubleshoot
> 1. **Reproducible Example**:
>    It looks like you've provided a good amount of detail, but to ensure we can reproduce the issue accurately, could you please share a minimum reproducible example? This includes a small subset of your dataset and the exact code you're using. You can refer to our guide on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).
> 2. **Package Versions**:
>    Ensure you are using the latest versions of `torch` and `ultralytics`. You can update your packages using the following commands:
>    ```shell
>    pip install --upgrade torch ultralytics
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    After updating, please try running your training script again to see if the issue persists.
> 3. **Training Configuration**:
>    Your training script looks mostly correct, but let's ensure a few things:
>    
>    * Verify that your dataset is correctly labeled and the `data.yaml` file paths are accurate.
>    * Ensure that the `single_cls=True` argument is correctly applied in your training command.
> 4. **Model Performance**:
>    Given that your model performs well on the training data but not on the test data, this might indicate overfitting. Here are a few suggestions to mitigate this:
>    
>    * **Increase Dataset Size**: If possible, add more diverse examples to your training dataset.
>    * **Data Augmentation**: Continue using data augmentation techniques, but ensure they are applied consistently. Consider using additional augmentations like random cropping, flipping, and scaling.
>    * **Regularization**: Implement techniques such as dropout or weight decay to help generalize better.
> 5. **Evaluation Metrics**:
>    Monitor additional metrics such as precision, recall, and mAP during training to get a better understanding of your model's performance. Tools like TensorBoard or wandb can be very helpful for this.
> 
> ### Example Code for Training with Augmentation
> Here's a refined example of how you might set up your training script with additional augmentations:
> 
> ```python
> import torch
> from ultralytics import YOLO
> 
> def train_model():
>     # Check if CUDA is available and set the device accordingly
>     device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
> 
>     # Load the YOLOv8 model and move it to the GPU if available
>     model = YOLO('yolov8n.pt')
>     model.to(device)
> 
>     # Train the model with additional augmentations
>     results = model.train(
>         data="data.yaml",
>         epochs=100,
>         imgsz=640,
>         single_cls=True,
>         augment=True,  # Enable data augmentation
>         hsv_h=0.015,  # Adjust hue
>         hsv_s=0.7,    # Adjust saturation
>         hsv_v=0.4,    # Adjust value
>         degrees=10.0, # Rotate images
>         translate=0.1, # Translate images
>         scale=0.5,    # Scale images
>         shear=0.0,    # Shear images
>         flipud=0.0,   # Flip images vertically
>         fliplr=0.5    # Flip images horizontally
>     )
> 
> if __name__ == '__main__':
>     train_model()
> ```
> 
> ### Next Steps
> * Please update your packages and try the refined training script.
> * Share a minimum reproducible example if the issue persists.
> * Monitor additional metrics during training to gain more insights.
> 
> Feel free to reach out if you have any further questions or need additional assistance. We're here to help! 😊

Thank you for the detailed answer. Appreciate it. 

I will update later on the results. Regarding image size, you don't think shrinking the images from their high quality to 360x640 might be a reason? If yes, can we train the model on original images or you don't recommend it?

**Comment by glenn-jocher** [comment]
@Avv22 you're welcome! 😊

Reducing the image size can indeed impact the model's performance, especially if the objects are small or detailed. Training on higher resolution images can sometimes yield better results, but it also requires more computational resources. 

### Training on Original Image Size
You can certainly train the model on the original image size (1080x1920). Here's how you can adjust your training script to use the original resolution:

```python
import torch
from ultralytics import YOLO

def train_model():
    # Check if CUDA is available and set the device accordingly
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    # Load the YOLOv8 model and move it to the GPU if available
    model = YOLO('yolov8n.pt')
    model.to(device)

    # Train the model with the original image size
    results = model.train(
        data="data.yaml",
        epochs=100,
        imgsz=(1080, 1920),  # Use original image size
        single_cls=True,
        augment=True,  # Enable data augmentation
        hsv_h=0.015,  # Adjust hue
        hsv_s=0.7,    # Adjust saturation
        hsv_v=0.4,    # Adjust value
        degrees=10.0, # Rotate images
        translate=0.1, # Translate images
        scale=0.5,    # Scale images
        shear=0.0,    # Shear images
        flipud=0.0,   # Flip images vertically
        fliplr=0.5    # Flip images horizontally
    )

if __name__ == '__main__':
    train_model()
```

### Considerations:
1. **GPU Memory**: Training on higher resolution images will require more GPU memory. Ensure your hardware can handle the increased load.
2. **Training Time**: Higher resolution images will increase training time. Be prepared for longer training sessions.
3. **Data Augmentation**: Continue using data augmentation to help the model generalize better.

### Next Steps:
- **Update Packages**: Ensure you are using the latest versions of `torch` and `ultralytics`.
- **Try the Updated Script**: Run the training script with the original image size.
- **Monitor Metrics**: Keep an eye on metrics like precision, recall, and mAP to evaluate performance.

Feel free to share your results or any further questions you might have. We're here to help! 🚀
